flight/PiOS/Common/pios_crossfire.c

```diff
@@ -0,0 +1,337 @@
+{
+	struct pios_crossfire_dev *dev = (struct pios_crossfire_dev *)context;
+
+	if (!PIOS_Crossfire_Validate(dev))
+		goto out_fail;
+
+	for (int i = 0; i < buf_len; i++) {
+		// Ignore any incoming stuff until we're expecting new data.
+		if(!dev->bytes_expected || dev->buf_pos >= CRSF_MAX_FRAMELEN) continue;
+
+		dev->u.rx_buf[dev->buf_pos++] = buf[i];
+		dev->bytes_expected--;
+
+		if(dev->buf_pos == CRSF_LEN_IDX)
+		{
+			// Read length field. Denotes payload, plus type field, plus CRC field.
+			dev->bytes_expected = dev->u.frame.length;
+
+			// Sanitize because we don't want to blow up when crap gets sent on serial.
+			if(dev->bytes_expected >= (CRSF_MAX_PAYLOAD + CRSF_TYPE_LEN + CRSF_CRC_LEN))
+				dev->bytes_expected = (CRSF_MAX_PAYLOAD + CRSF_TYPE_LEN + CRSF_CRC_LEN);
+		}
+
+		if (!dev->bytes_expected)
+		{
+			// Frame complete, decode.
+			PIOS_Crossfire_UnpackFrame(dev);
+			PIOS_RCVR_ActiveFromISR();
+		}
+	}
+
+	dev->rx_timer = 0;
+
+	// Keep pumping data.
+	*headroom = CRSF_MAX_FRAMELEN;
+	*task_woken = false;
+	return buf_len;
+
+out_fail:
+	*headroom = 0;
+	*task_woken = false;
+	return 0;
+}
+
+static void PIOS_Crossfire_ResetBuffer(struct pios_crossfire_dev *dev)
+{
+	dev->buf_pos = 0;
+	dev->bytes_expected = CRSF_MAX_FRAMELEN;
+}
+
+static void PIOS_Crossfire_UnpackFrame(struct pios_crossfire_dev *dev)
+{
+	// Pull CRC from type field and payload.
+	uint8_t crc = PIOS_CRC_updateCRC_TBS(0, &dev->u.frame.type, dev->u.frame.length - CRSF_CRC_LEN);
+
+	if(crc == CRSF_CRCFIELD(dev->u.frame))
+	{
+		// Apparently only RC channel payload gets sent for now.
+		if(dev->u.frame.type == CRSF_FRAME_RCCHANNELS)
+		{
+			// From pios_sbus.c
+			uint8_t *s = dev->u.frame.payload;
+			uint16_t *d = dev->channel_data;
+
+			#define F(v,s) (((v) >> (s)) & 0x7ff)
+
+			/* unroll channels 1-8 */
+			*d++ = F(s[0] | s[1] << 8, 0);
+			*d++ = F(s[1] | s[2] << 8, 3);
+			*d++ = F(s[2] | s[3] << 8 | s[4] << 16, 6);
+			*d++ = F(s[4] | s[5] << 8, 1);
+			*d++ = F(s[5] | s[6] << 8, 4);
+			*d++ = F(s[6] | s[7] << 8 | s[8] << 16, 7);
+			*d++ = F(s[8] | s[9] << 8, 2);
+			*d++ = F(s[9] | s[10] << 8, 5);
+
+			/* unroll channels 9-16 */
+			*d++ = F(s[11] | s[12] << 8, 0);
+			*d++ = F(s[12] | s[13] << 8, 3);
+			*d++ = F(s[13] | s[14] << 8 | s[15] << 16, 6);
+			*d++ = F(s[15] | s[16] << 8, 1);
+			*d++ = F(s[16] | s[17] << 8, 4);
+			*d++ = F(s[17] | s[18] << 8 | s[19] << 16, 7);
+			*d++ = F(s[19] | s[20] << 8, 2);
+			*d++ = F(s[20] | s[21] << 8, 5);
+		}
+
+		// Consider link proper anyway, even if the packet
+		// didn't deliver RC data.
+		dev->failsafe_timer = 0;
+	}
+
+	PIOS_Crossfire_ResetBuffer(dev);
+}
+
+static void PIOS_Crossfire_Supervisor(uintptr_t context)
+{
+	struct pios_crossfire_dev *dev = (struct pios_crossfire_dev *)context;
+	PIOS_Assert(PIOS_Crossfire_Validate(dev));
+
+	// If more than 1.6ms passed without communication, safe to say that a new
+	// packet is inbound.
+	if (++dev->rx_timer > 1)
```

Comment:
I assume this is the best that can be gotten away with here..  It's unfortunate, as it makes resync difficult if things are a byte off, etc.